File: android/codec/src/test/kotlin/com/farosos/codec/TestVectorFile.kt
package com.farosos.codec

import org.json.JSONObject
import java.io.File

/**
 * Utilidades compartidas para leer `spec/test-vectors.json` desde los tests
 * de `:codec` (layout legado, Caso B, autenticación) — evita repetir el
 * mismo helper de resolución de ruta + parseo de hex en cada archivo de
 * vectores.
 */
object TestVectorFile {
    fun load(): JSONObject {
        val vectorsFile = File(repoRootDir(), "spec/test-vectors.json")
        return JSONObject(vectorsFile.readText())
    }

    fun hexToBytes(hex: String): ByteArray =
        ByteArray(hex.length / 2) { i -> hex.substring(i * 2, i * 2 + 2).toInt(16).toByte() }

    fun bytesToHex(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it) }

    fun hexByte(hex: String): Int = hex.removePrefix("0x").toInt(16)

    // Gradle corre los tests del módulo `:codec` con working dir = android/codec/
    private fun repoRootDir(): File = File(System.getProperty("user.dir"), "../..").canonicalFile
}

File: ios/Tests/PacketCodecTests/TestVectorFile.swift
import XCTest

/// Utilidades compartidas para leer `spec/test-vectors.json` desde los
/// tests de `PacketCodecTests` (layout legado, Caso B, autenticación) —
/// evita repetir el mismo helper de resolución de ruta + parseo de hex en
/// cada archivo de vectores.
enum TestVectorFile {
    static func load() throws -> [String: Any] {
        let vectorsURL = repoRootURL()
            .appendingPathComponent("spec")
            .appendingPathComponent("test-vectors.json")
        let data = try Data(contentsOf: vectorsURL)
        return try XCTUnwrap(JSONSerialization.jsonObject(with: data) as? [String: Any])
    }

    static func hexToData(_ hex: String) -> Data {
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            data.append(UInt8(hex[index..<next], radix: 16)!)
            index = next
        }
        return data
    }

    static func hexByte(_ hex: String) -> UInt8 {
        let digits = hex.hasPrefix("0x") ? String(hex.dropFirst(2)) : hex
        return UInt8(digits, radix: 16)!
    }

    private static func repoRootURL() -> URL {
        // #filePath = .../ios/Tests/PacketCodecTests/TestVectorFile.swift
        URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent() // quita TestVectorFile.swift -> PacketCodecTests/
            .deletingLastPathComponent() // -> Tests/
            .deletingLastPathComponent() // -> ios/
            .deletingLastPathComponent() // -> raíz del repo
    }
}
